### DISCUSSION TITLE
How should we handle orphaned HelmRelease when changing targetNamespace?

### DISCUSSION THREAD

**Discussion by scottrigby** [fluxcd/flux2#1691] [Q&A]
Kicking off a process for community discussion on how we should handle the use case in this issue https://github.com/fluxcd/helm-controller/issues/237 which was raised during today's bug scrub 🐛 

AFAIK there is not currently a straightforward answer to this question:
> How should we handle orphaned HelmRelease when changing targetNamespace?

My understanding is Flux Helm Controller `targetNamespace` (`flux create helmrelease [name] --target-namespace [string]` installs the helm-managed resources into a different namespace than the release metadata. And this is intended for use cases where you specifically want this non-standard behavior, as one of several access control strategies for GitOps. This is not something the Helm client supports, therefore I'm thinking we may need more feedback from the Flux Helm Controller end user community on how to follow the [principle of least astonishment](https://en.wikipedia.org/wiki/Principle_of_least_astonishment) when this feature is opted-into.

## 📊 Poll for end user feedback

Should Helm Controller:

1. Do the equivalent of `helm uninstall` the release in NS no longer specified in the `HelmRelease` CRD declaration, and `helm update --install` the release into the newly specified NS?
1. Do the equivalent of `helm update --install` the release into the newly specified NS, but leave the previously defined release in the old NS? Essentially "orphaning" the old release
    - In this case, what level of alerting is sufficient? Negative polarity / abnormal = true condition? Warning event (via Notification Controller, etc)?
1. Should this operation not be permitted without human intervention, if there is no way to know how a user wants to automate it (error out)?
1. Are there other options?

**Comment by hiddeco** [comment]
The pattern we have seen for cases like this, where the desired behavior may differ per user, is to introduce action defining policies (and/or modes). The Kubernetes core has this same concept, for example for `PersistentVolume` resources: https://kubernetes.io/docs/concepts/storage/persistent-volumes/#reclaim-policy.

As I noted in https://github.com/fluxcd/helm-controller/issues/237#issuecomment-802771718, orphaned releases at the moment do not just happen for `targetNamespace` changes, but other name-related changes as well. In addition, there are constraints around renaming and/or moving releases, of which one I mentioned in the comment. Also note that renaming a release itself is because of this marked as `wontfix` for now by the Helm project itself: https://github.com/helm/helm/issues/1809#issuecomment-467985478

Something that should be possible without too much fuzz is changing (or rather migrating) objects on a `storageNamespace` change, which would likely very much be a copy operation to the other namespace with garbage collection afterwards.

Given the above, all 3 options could be available (and more, if options come to light that seem reasonable). As of default behavior, I believe that as with `PersistentVolume`, going with the safest but least annoying option is the best.
